BootloaderCommonPkg/Library/ShellLib/CmdMem.c

```diff
@@ -50,115 +52,187 @@ ShellCommandMemFunc (
+          Index++;
+        } else {
+          ShellPrint (L"Error, '-w' requires [1|2|4|8]\n");
+        }
+      } else if (StrCmp (Argv[Index], L"-c") == 0) {
+        if ((Index + 1) < Argc) {
+          Count = StrHexToUintn (Argv[Index+1]);
+          Index++;
+        } else {
+          ShellPrint (L"Error, '-c' requires a value\n");
+        }
+      } else if (StrCmp (Argv[Index], L"-pci") == 0) {
+        IsPciAddr = TRUE;
+      } else if (StrCmp (Argv[Index], L"-io") == 0) {
+        IsIoAddr = TRUE;
+      } else if (Write == FALSE) {
+        Value = StrHexToUintn (Argv[Index]);
+        Write = TRUE;
+      } else {
+        ShellPrint (L"Invalid parameter '%s'\n", Argv[Index]);
+        Status = EFI_INVALID_PARAMETER;
+      }
+    }
+  } else if (Argc >= 9) {
+    ShellPrint (L"Error, too many arguments\n");
+    Status = EFI_INVALID_PARAMETER;
   } else {
-    // Invalid
-    goto usage;
+    Status = EFI_INVALID_PARAMETER;
   }
 
-  // Parse access width
-  if (Argc > 2) {
-    Width = StrHexToUintn (Argv[2]);
+  while (!EFI_ERROR (Status)) {
     if ((Width != 1) && (Width != 2) && (Width != 4) && (Width != 8)) {
-      ShellPrint (L"Error: Width must be 1, 2, 4, or 8\n");
-      return EFI_INVALID_PARAMETER;
+      ShellPrint (L"Error, width %u not supported\n", Width);
+      Status = EFI_INVALID_PARAMETER;
+      break;
     }
-  }
 
-  // Parse access address
-  Addr = StrHexToUintn (Argv[1]);
-
-  // Make sure access is aligned
-  if ((Addr & (Width - 1)) != 0) {
-    ShellPrint (L"Error: Unaligned access\n");
-    return EFI_INVALID_PARAMETER;
-  }
-
-  if (Write) {
-    switch (Width) {
-    case 1:
-      * (volatile UINT8 *)Addr = (UINT8 )Value;
-      break;
-    case 2:
-      * (volatile UINT16 *)Addr = (UINT16)Value;
-      break;
-    case 4:
-      * (volatile UINT32 *)Addr = (UINT32)Value;
-      break;
-    case 8:
-      * (volatile UINT64 *)Addr = (UINT64)Value;
-      break;
-    default:
+    if (IsIoAddr && IsPciAddr) {
+      ShellPrint (L"Error, cannot be both -pci & -io\n");
+      Status = EFI_INVALID_PARAMETER;
       break;
+    } else if (IsPciAddr) {
+      Addr = (UINTN) (PcdGet64 (PcdPciExpressBaseAddress) +
+        PCI_EXPRESS_LIB_ADDRESS ((Addr >> 24) & 0xFF, (Addr >> 16) & 0xFF, (Addr >> 8) & 0xFF, Addr & 0xFF));
     }
-  } else {
-    // Pad backward to 16-byte boundary
-    ItemsPerRow  = 16 / Width;
-    Start        = (Addr & 0xf) / Width;
-    Count       += Start;
-
-    if (Start > 0) {
-      ShellPrint (L"%08x: ", Addr & ~0xf);
-      for (Index = 0; Index < Start * (Width * 2 + 1); Index++) {
-        ShellPrint (L" ");
-      }
+
+    // Make sure access is aligned
+    if ((Addr & (Width - 1)) != 0) {
+      ShellPrint (L"Error, unaligned access\n");
+      Status = EFI_INVALID_PARAMETER;
+      break;
     }
 
-    for (Index = Start; Index < Count; Index++) {
-      if ((Index % ItemsPerRow) == 0) {
-        ShellPrint (L"%08x: ", Addr);
+    if (Write) {
```

Comment:
Maybe better to break the code below into Read and Write sub-functions to make it easier to read.